tree/branch_node.go

```diff
@@ -0,0 +1,158 @@
+package tree
+
+import "fmt"
+
+// BranchNode represents a Node with an array of nodes and a sharedAddress
+// sharedAddress - part of the address that it links to
+// nodes are links to other nodes, which their addresses are suffixed by sharedAddress
+// [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, f ]
+//     [ x ] -> LeafNode
+//     [ not found ] -> EmptyNode
+//
+type BranchNode struct {
+	nodes         map[Unit]Node // TODO this can probably be an array if it help ?
+	sharedAddress []Unit
+	parent        Node
+	hash          []byte
+}
+
+// NewBranchNode returns a new BranchNode
+func NewBranchNode(sharedAddress []Unit, parent Node) Node {
+	return &BranchNode{
+		sharedAddress: sharedAddress,
+		nodes:         map[Unit]Node{},
+		parent:        parent,
+	}
+}
+
+// GetChild returns a child node
+// returns Node - one of it's children
+// if there isn't one
+// it returns EmptyNode - no node in this position
+//
+func (b *BranchNode) GetChild(key []Unit) Node {
+	// key - ABCDE
+	// b.sharedAddress - AC
+	// SharedPrefix() - A != b.sharedAddress
+	// if the node CAN'T exist in this prefix return an EmptyNode
+	if !EqualUnits(SharedPrefix(b.sharedAddress, key), b.sharedAddress) {
```

Comment:
If `b.sharedAddress` == `key`, couldn't this cause an index OOB error in `EqualUnits` since `len(SharedPrefix(b.sharedAddress, key)` == `len(b.sharedAddress)`